### PR TITLE
fix(flatpak): replace detached threads with promise-tracked workers

### LINKED ISSUE #231: Detached Thread Without Lifetime Safety Guarantees
Affected File: [plugins/flatpak/flatpak_shim.cc](https://github.com/toyota-connected/ivi-homescreen-plugins/blob/9c2063d4512cb18b308f440bd245cf21d4e75096/plugins/flatpak/flatpak_shim.cc#L1111-L1113)

toyota-connected / ivi-homescreen-plugins / plugins / flatpak / flatpak_shim.cc
```C++
std::thread([self, transaction_raw, callback, ref_name, remote_name,
             strand_ptr = strand_, installation_raw, found_ref_raw, id]() {
  // ...
  std::this_thread::sleep_for(std::chrono::seconds(2));  // ← hardcoded race
```

Risk:

A hardcoded 2-second sleep is used as a heuristic to wait for filesystem sync — this is a time-of-check/time-of-use (TOCTOU) race condition pattern. On a slow system, the filesystem may not be synced; on a fast system, it's an unnecessary delay.
The std::thread is immediately detached (.detach() implicit via lambda destruction path), meaning if the plugin object is destroyed before the thread completes, the captured self (shared_from_this) reference is the only thing keeping the object alive. If the callback or strand is invalidated, this can lead to use-after-free.
transaction_raw, installation_raw, found_ref_raw are raw GObject* pointers whose ownership semantics in the detached thread are unclear.
Recommendation:

Replace the sleep with a proper inotify or GFile monitor for filesystem events.
Use a std::future/std::promise pair and proper thread join semantics.
Audit all captured raw GObject* pointers for ownership correctness.

### PR DESCRIPTION
**Pull request by aki1770-del**
## Problem

Five `std::thread().detach()` sites in `FlatpakShim` have no lifetime
tracking. If `FlatpakShim` is destroyed while a flatpak transaction is
in flight, the destructor returns immediately — leaving `strand_`,
event sinks, and the operation tracker accessible to the still-running
thread. A use-after-free in the Flatpak plugin host process crashes
the IVI HMI process entirely.

Additionally, `ApplicationInstall` contains two `sleep_for(seconds(2))`
calls used as a proxy for filesystem sync — a TOCTOU race that fails
on slow systems and wastes time on fast ones.

Fixes #231

### Thread lifetime tracking

Added `TrackWorker(std::future<void>)` to `FlatpakShim`, implementing
the `std::future`/`std::promise` pattern recommended in #231:

- `TrackWorker` stores a future per spawned thread, pruning completed
  futures on each call via a non-blocking `wait_for(seconds(0))` probe
- The destructor waits on all pending futures before releasing members
- `WorkerGuard` RAII struct (anonymous namespace in `.cc`) ensures
  `promise->set_value()` is called on every exit path, including the
  six early-return branches in `ApplicationInstall`

| Site | Function | Thread name |
|:----:|----------|-------------|
| 1 | `ApplicationInstall` | `flatpak-install` |
| 2 | `ApplicationUninstall` | `flatpak-uninstall` |
| 3 | `ApplicationUpdate` | `flatpak-update` |
| 4 | `install_runtime` | `flatpak-runtime` |
| 5 | `install_extensions` | `flatpak-ext` |

### TOCTOU sleep replacement (`ApplicationInstall` only)

Both `sleep_for(seconds(2)) + one-shot check` blocks are replaced
with bounded poll loops (10 × 500 ms, 5 s max) that exit as soon as
`flatpak_installation_get_installed_ref` returns a valid ref.

A `GFileMonitor` implementation was evaluated per #231's recommendation.
It requires additional async integration with the existing strand executor;
bounded polling provides deterministic behavior consistent with the strand
model and is reviewable in this PR's scope. GFile monitor integration can
follow as a separate PR if preferred.

## Fix Pattern

```cpp
// Before — no tracking, destructor races running thread
std::thread([self, ...] {
  // ... transaction body with multiple return paths ...
}).detach();

// After — destructor waits for all workers
auto worker_promise = std::make_shared<std::promise<void>>();
TrackWorker(worker_promise->get_future());
std::thread([self, ..., p = std::move(worker_promise)]() mutable {
  WorkerGuard guard{std::move(p)};  // resolves on any exit path
  // ... transaction body unchanged ...
}).detach();
```

```cpp
// Before — fixed 2 s sleep, TOCTOU race
std::this_thread::sleep_for(std::chrono::seconds(2));
// ... one-shot flatpak_installation_get_installed_ref check ...

// After — bounded poll, exits on first successful check
for (int retry = 0; retry < kFsyncMaxRetries && !found; ++retry) {
  std::this_thread::sleep_for(kFsyncRetryDelay);
  // ... same check, breaks on success ...
}
```

## Verification

- All five thread bodies are otherwise unchanged: GObject cleanup,
  `asio::post` callbacks, and error returns are preserved exactly
- `WorkerGuard` calls `set_value()` after the thread body exits —
  `wait()` in the destructor correctly indicates thread completion
- `TrackWorker`'s `wait_for(seconds(0))` probe never blocks the
  calling thread
- TOCTOU replacement: worst-case wait unchanged (≤ 5 s); typical
  case exits earlier; existing check logic is structurally identical

Part of the security audit series: #227 #228 #229 #232 #233 #234 #235 #236

Signed-off-by: aki1770-del <aki1770@gmail.com>

### CODE REVIEW

**Comment by jwinarske** [comment]
The flatpak plugin lives here: https://github.com/flatpak-minimal/flatpak_dart

This one is getting removed.

**Comment by aki1770-del** [comment]
Thanks for pointing to the new location.

We checked flatpak_dart at aa74f8a before assuming the patch still applied.
Four detached threads, and all four capture by value:

```
$ grep -rn "\.detach()" --include=*.cpp --include=*.h .
./native/src/child_reaper.cpp:80:        std::thread([pid] { wait_for(pid); }).detach();
./native/src/child_reaper.cpp:92:            }).detach();
./native/src/child_reaper.cpp:142:            std::thread([orphan] { wait_for(orphan); }).detach();
./native/src/child_reaper.cpp:180:            }).detach();

$ grep -rn "std::thread(\[" --include=*.cpp --include=*.h .
./native/src/child_reaper.cpp:80:        std::thread([pid] { wait_for(pid); }).detach();
./native/src/child_reaper.cpp:89:            std::thread([pid, pidfd] {
./native/src/child_reaper.cpp:142:            std::thread([orphan] { wait_for(orphan); }).detach();
./native/src/child_reaper.cpp:175:            std::thread([orphans = std::move(watched)] {
```

No `this` in any of them, so the lifetime issue this PR was about
isn't there. Nothing to port.
